api/observation.go

```diff
@@ -0,0 +1,389 @@
+	"github.com/ONSdigital/go-ns/common"
+	"github.com/ONSdigital/log.go/log"
+	"github.com/gorilla/mux"
+	"github.com/pkg/errors"
+)
+
+// Upper limit, if this is not big enough, we may need to consider increasing value
+// and then if this has a performance hit then consider paging
+const (
+	defaultOffset = 0
+
+	getObservationsAction = "getObservations"
+)
+
+var (
+	observationNotFound = map[error]bool{
+		errs.ErrDatasetNotFound:      true,
+		errs.ErrEditionNotFound:      true,
+		errs.ErrVersionNotFound:      true,
+		errs.ErrObservationsNotFound: true,
+	}
+
+	observationBadRequest = map[error]bool{
+		errs.ErrTooManyWildcards: true,
+	}
+)
+
+func (api *API) getObservations(w http.ResponseWriter, r *http.Request) {
+	ctx := r.Context()
+	vars := mux.Vars(r)
+	datasetID := vars["dataset_id"]
+	edition := vars["edition"]
+	version := vars["version"]
+
+	// TODO call audit (attempt) once it has its own library
+	logData := log.Data{"dataset_id": datasetID, "edition": edition, "version": version}
+
+	observationsDoc, err := api.doGetObservations(ctx, datasetID, edition, version, r, logData)
+	if err != nil {
+		// TODO call audit (unsuccessful) once it has its own library
+		handleObservationsErrorType(ctx, w, err, logData)
+		return
+	}
+
+	// TODO call audit (successful) once it has its own library
+
+	setJSONContentType(w)
+
+	// The ampersand "&" is escaped to "\u0026" to keep some browsers from
+	// misinterpreting JSON output as HTML. This escaping can be disabled using
+	// an Encoder that had SetEscapeHTML(false) called on it.
+	enc := json.NewEncoder(w)
+	enc.SetEscapeHTML(false)
+
+	if err = enc.Encode(observationsDoc); err != nil {
+		handleObservationsErrorType(ctx, w, errors.WithMessage(err, "failed to marshal metadata resource into bytes"), logData)
+		return
+	}
+
+	log.Event(ctx, "get observations endpoint: successfully retrieved observations relative to a selected set of dimension options for a version", log.INFO, logData)
+}
+
+func (api *API) doGetObservations(ctx context.Context, datasetID, edition, version string, r *http.Request, logData log.Data) (*models.ObservationsDoc, error) {
+
+	// TODO implement auth once the auth features are moved to their own library
+	authorised := api.authenticate(r, logData)
+	userAuthToken := getUserAuthToken(r.Context())
+
+	// Get dataset from dataset API
+	datasetDoc, err := api.datasetClient.Get(ctx, userAuthToken, api.cfg.ServiceAuthToken, "", datasetID)
+	if err != nil {
+		log.Event(ctx, "get observations: dataset api get /datasets/{dataset_id} returned an error", log.ERROR, log.Error(err), logData)
+		return nil, err
+	}
+
+	// If not authorised, only published datasets are accessible
+	if !authorised {
+		if datasetDoc.State != dataset.StatePublished.String() {
+			logData["dataset_doc"] = datasetDoc
+			log.Event(ctx, "get observations: found no published dataset", log.ERROR, log.Error(errs.ErrDatasetNotFound), logData)
+			return nil, errs.ErrDatasetNotFound
+		}
+	}
```

Comment:
I think this feels like its isolated _unit of work_ that could probably be moved into a separate function. This code would be the same for any scenario where get a dataset doc - if we move it to its own function we can reuse it instead of duplicating that logic many times.